### ISSUE TITLE
NixOS: Confusing "FileNotFound" error with linking libc/libc++ for `linux-gnu`

### ISSUE THREAD

**Issue by sagehane** [ziglang/zig#10602]
### Zig Version

0.10.0-dev.290+3901b6fb0

### Steps to Reproduce

Try running a program that links to `libc` or `libc++`, targetting `x86_64-linux-gnu` on NixOS.

Sample program:
```c
#include<stdio.h>

int main(void) {
    printf("Hello World\n");
    return 0;
}
```

### Expected Behavior

I would expect the program to run properly, or have a consistent error message for linking `libc` and `libc++`.

The program works if `--dynamic-linker /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2` is passed:
```
$ zig run -lc++ hello.c -target x86_64-linux-gnu --dynamic-linker /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2
Hello World
```

### Actual Behavior

```sh-session
$ zig run -lc hello.c -target x86_64-linux-musl
Hello World

$ zig run -lc hello.c -target x86_64-linux-gnu
warning: the host system does not appear to be capable of executing binaries from the target because the host dynamic linker is '/nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2', while the target dynamic linker is '/lib64/ld-linux-x86-64.so.2'. Consider using '--dynamic-linker'
error: the following command failed to execve with 'FileNotFound':
/home/plumeus/.cache/zig/o/f1117a11db5928d0a448dcf473c855e6/hello

$ zig run -lc++ hello.c -target x86_64-linux-gnu
error: the following command failed to execve with 'FileNotFound':
/home/plumeus/.cache/zig/o/c8a055afde64d1322cc9140f3671fa1d/hello
```

**Comment by sagehane** [comment]
This is reproducible on 0.8.1 and 0.9.0 too, although the "warning" for the `-lc` isn't available on 0.8.1.

---

Hopefully relevant strace outputs on Zig master:
```sh-session
$ strace zig run -lc hello.c -target x86_64-linux-gnu

<lots of lines omitted>

openat(AT_FDCWD, "/nix/store/rf9d67pk49l2sq26qiwwp1wl727qf11f-acl-2.3.1/lib", O_RDONLY|O_LARGEFILE|O_CLOEXEC|O_PATH|O_DIRECTORY) = 10
readlinkat(10, "libc.so.6", 0x7ffd7e3ba050, 4096) = -1 ENOENT (No such file or directory)
close(10)                               = 0
openat(AT_FDCWD, "/nix/store/5b40cks3m9nhkxiy4g9jsaqbp1vipc6g-attr-2.5.1/lib", O_RDONLY|O_LARGEFILE|O_CLOEXEC|O_PATH|O_DIRECTORY) = 10
readlinkat(10, "libc.so.6", 0x7ffd7e3ba050, 4096) = -1 ENOENT (No such file or directory)
close(10)                               = 0
openat(AT_FDCWD, "/nix/store/s9qbqh7gzacs7h68b2jfmn9l6q4jwfjz-glibc-2.33-59/lib", O_RDONLY|O_LARGEFILE|O_CLOEXEC|O_PATH|O_DIRECTORY) = 10
readlinkat(10, "libc.so.6", "libc-2.33.so", 4096) = 12
close(10)                               = 0
close(8)                                = 0
faccessat(AT_FDCWD, "/lib64/ld-linux-x86-64.so.2", F_OK) = -1 ENOENT (No such file or directory)
write(2, "warning: the host system does no"..., 129warning: the host system does not appear to be capable of executing binaries from the target because the host dynamic linker is ') = 129
write(2, "/nix/store/s9qbqh7gzacs7h68b2jfm"..., 82/nix/store/s9qbqh7gzacs7h68b2jfmn9l6q4jwfjz-glibc-2.33-59/lib/ld-linux-x86-64.so.2) = 82
write(2, "', while the target dynamic link"..., 39', while the target dynamic linker is ') = 39
write(2, "/lib64/ld-linux-x86-64.so.2", 27/lib64/ld-linux-x86-64.so.2) = 27
write(2, "'. Consider using '--dynamic-lin"..., 36'. Consider using '--dynamic-linker') = 36
write(2, "\n", 1
)                       = 1
write(2, "error: the following command fai"..., 52error: the following command failed to execve with ') = 52
write(2, "FileNotFound", 12FileNotFound)            = 12
write(2, "':\n", 3':
)                     = 3
write(2, "/home/plumeus/.cache/zig/o/1a54d"..., 65/home/plumeus/.cache/zig/o/1a54da14ca00a74582296b11daa3d9c6/hello) = 65
write(2, "\n", 1
)                       = 1
```

Is it supposed to be trying to use the native libc for this?

**Comment by sagehane** [comment]
Tried to see if I could reproduce this on Arch Linux, and the command ran just fine. So this definitely seems like a NixOS-specific issue.

Posting ldd output here just in case.
```sh-session
$ zig build-exe -lc++ hello.c -target x86_64-linux-gnu

$ ldd hello
	linux-vdso.so.1 (0x00007fffd42ba000)
	libc.so.6 => /nix/store/s9qbqh7gzacs7h68b2jfmn9l6q4jwfjz-glibc-2.33-59/lib/libc.so.6 (0x00007f1df61df000)
	/lib64/ld-linux-x86-64.so.2 => /nix/store/s9qbqh7gzacs7h68b2jfmn9l6q4jwfjz-glibc-2.33-59/lib64/ld-linux-x86-64.so.2 (0x00007f1df63a6000)
```

**Comment by andrewrk** [comment]
Everything is working as designed, and in fact it's working smoothly and beautifully. If you want the binary to run natively then use `-target native-native-gnu` instead of `-target x86_64-linux-gnu`.

If you want to create a cross compile that works on NixOS then pass the appropriate `--dynamic-linker` path for the system you want to run the binary on.

The error message that you pasted above explains the problem perfectly:

> warning: the host system does not appear to be capable of executing binaries from the target because the host dynamic linker is '/nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2', while the target dynamic linker is '/lib64/ld-linux-x86-64.so.2'. Consider using '--dynamic-linker'

**Comment by andrewrk** [comment]
Missing warning message for the -lc++ case fixed here: 557a097523a8c30c25579db3c634c09c9979d3a2
